### PR TITLE
feat(inference): add trusted llama.cpp publication evidence

### LINKED ISSUE #8250: [Managed inference] Add trusted llama.cpp image publication evidence
## Summary

Add the next delivery slice under #8231 and accepted epic #8144: trusted, digest-first publication and supply-chain evidence for the NemoClaw-owned llama.cpp server image.

PR #8235 established the declarative image contract and secretless native amd64/arm64 PR builds. This slice adds the trusted publication machinery while keeping publication opt-in until the required GHCR package settings and protected DGX Spark runner are provisioned. It must not add an always-on `main` gate that cannot run in the current repository environment.

## Observable outcome

- `managed-inference/images/llama-cpp/image.yaml` is the sole source of publication policy and declares the repository, run-unique candidate naming, platforms, evidence formats, signing identity, vulnerability policy, anonymous-pull requirement, receipt schema, and enablement boundary.
- The compiler validates and exports that policy; no TypeScript serving or publication defaults are introduced.
- A trusted workflow can publish exact native platform digests, assemble an index from those descriptors, verify the required evidence, and emit one machine-readable receipt.
- No mutable consumer alias moves in this slice.
- Pull requests remain secretless and perform no registry writes.

## Scope

- Extend the dedicated llama.cpp image workflow rather than coupling it to managed-agent image cohort promotion.
- Publish native amd64 and arm64 artifacts by digest first and assemble the candidate OCI index only from the exact validated descriptors.
- Generate and verify SPDX JSON SBOMs and SLSA provenance for the published artifacts.
- Keylessly sign the candidate index with GitHub OIDC and verify the expected workflow identity, issuer, and transparency-log inclusion.
- Scan exact digests with a declarative `high` / `onlyFixed` vulnerability policy.
- Verify anonymous exact-digest pulls.
- Bind repository, workflow run, source revision, CUDA base digests, platform descriptors, index digest, SBOM, provenance, signature, vulnerability result, and anonymous-pull result into a versioned receipt.
- Add contract tests and hostile fixtures that fail closed on missing, mutable, mismatched, unsigned, unscanned, or unbound evidence.

## Enablement boundary

- Keep the trusted publication path manually gated or disabled by declarative YAML until repository administrators provision the GHCR package permissions/settings and protected DGX Spark execution environment required by #8231.
- Do not silently weaken or skip evidence when prerequisites are unavailable.
- Do not add a required `main` check that is known to be unrunnable.
- Use `GITHUB_TOKEN` and GitHub OIDC; do not add a static signing or registry secret.

## Acceptance criteria

- Exact publication inputs and policy live in YAML and are validated before any registry write.
- Workflow permissions are least-privilege and publication jobs cannot run from pull requests or untrusted refs.
- Platform and index digests are assembled and verified without mutable consumer tags.
- SPDX, SLSA, signature, vulnerability, anonymous-pull, and descriptor bindings are covered by deterministic tests, including hostile mutations.
- A canonical receipt can be generated and re-verified from exact digests.
- Existing vLLM selection, llama.cpp explicit-only recipe behavior, onboarding, and provider registration remain unchanged.
- The publication path remains safe to merge before external runner/package provisioning.

## Following slice

Run the protected DGX Spark qualification against the exact candidate digest, prove full NVIDIA GPU offload plus health/completion, then pin the successful owned index digest into the declarative recipe.

## References

- Parent delivery issue: #8231
- Accepted epic: #8144
- Image build slice: #8235
- NVIDIA DGX Spark playbooks: https://github.com/NVIDIA/dgx-spark-playbooks
- NVIDIA AICR publication precedent: https://github.com/NVIDIA/aicr/blob/b8d410fafea6a0bba8d4144947b9da1df831bbd8/.github/workflows/on-tag.yaml
- NVIDIA AICR attestation precedent: https://github.com/NVIDIA/aicr/blob/b8d410fafea6a0bba8d4144947b9da1df831bbd8/.github/workflows/attest-images.yaml

Part of #8231 and #8144.

### PR DESCRIPTION
**Pull request by ericksoa**
## Summary

This change adds a digest-first, manually gated publication path for the NemoClaw-owned llama.cpp server image. Publication remains disabled in declarative YAML until the public GHCR package and protected DGX Spark qualification infrastructure exist, and this slice does not move a consumer alias.

## Related Issue

Fixes #8250

## Changes

- Declare the publication repository, platforms, evidence policy, signing identity, vulnerability threshold, receipt schema, and enablement boundary in `managed-inference/images/llama-cpp/image.yaml`.
- Validate and export the exact publication contract from the existing manifest compiler. The dedicated llama.cpp publisher consumes this configuration; workflow-local defaults would violate the accepted declarative serving-configuration decision.
- Add a main-only manual publisher that builds native digests, requires a public GHCR package before registry writes, assembles an exact two-platform index, scans both platform digests, and creates no consumer alias.
- Add an independently gated reusable workflow for SPDX SBOM attestations, SLSA provenance, and Sigstore keyless index signing.
- Add a fail-closed receipt verifier that binds source, CUDA bases, descriptors, attestations, scans, anonymous per-platform pulls, and immutable local image IDs.
- Add workflow-contract and hostile-mutation tests for the compiler, trust boundary, digest handoff, and evidence verifier.

## Type of Change

- [x] Code change (feature, bug fix, or refactor)
- [ ] Code change with doc updates
- [ ] Doc only (prose changes, no code sample modifications)
- [ ] Doc only (includes code sample changes)

## Quality Gates

- [x] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [ ] Docs updated for user-facing behavior changes
- [x] Docs not applicable — justification: Publication remains disabled; no recipe digest, consumer alias, onboarding default, operator procedure, or supported surface changes.
- [x] Sensitive paths changed (security, policy, credentials, preflight, onboarding, inference, runner, sandbox, or messaging)
- [x] Sensitive-path review completed or maintainer-approved waiver recorded — reviewer/approval link/justification: Independent supply-chain review of the complete change through `c3fc700f6` found no remaining trust-boundary, evidence-integrity, or workflow-permission blocker.
- [ ] Non-success, skipped, or missing CI check accepted by maintainer — check name, approval link, and follow-up issue:

## Documentation Writer Review

- [x] Documentation writer subagent reviewed the completed changes
- Result: `no-docs-needed`
- Evidence: Publication is declaratively disabled, protected DGX Spark inputs are unset, and this change adds no consumer alias, recipe digest, onboarding default, operator procedure, or supported surface.
- Agent: Codex Desktop

## DGX Station Hardware Evidence

- [ ] Tested on DGX Station
- Tested commit:
- Station profile/scenario:
- Result:
- Supporting evidence:

## Verification

- [x] PR description includes a `Signed-off-by:` line and every commit appears as `Verified` in GitHub
- [x] Normal `pre-commit`, `commit-msg`, and `pre-push` hooks passed, or `npm run validate:pr` passed after refreshing `origin/main` when hooks were skipped or unavailable
- [x] Targeted behavior tests pass for the current change set, or tests are marked not applicable above — command/result or justification: 66 tests passed across the four focused integration files, including the dependency-review guard.
- [ ] Applicable broad gate passed — `npm test` for broad runtime/test-harness changes; `npm run check` for repo-wide validation/coverage changes — command/result:
- [x] Quality Gates section completed with required justifications or waivers
- [x] No secrets, API keys, or credentials committed
- [ ] `npm run docs` builds without warnings (doc changes only)
- [ ] Doc pages follow the [style guide](https://github.com/NVIDIA/NemoClaw/blob/main/docs/CONTRIBUTING.md) (doc changes only)
- [ ] New doc pages include SPDX header and frontmatter (new pages only)

---
Signed-off-by: Aaron Erickson <aerickson@nvidia.com>

### CODE REVIEW

File: test/llama-cpp-image-workflow.test.ts [hunk 1]
```diff
@@ -154,6 +198,182 @@ describe("llama.cpp image PR workflow", () => {
...
+    expect(crypto.run).toContain("cosign verify \\");
+    expect(crypto.run).toContain("cosign verify-attestation \\");
+    expect(crypto.run).toContain('--certificate-identity "$CERTIFICATE_IDENTITY"');
+    expect(crypto.run).toContain('--certificate-oidc-issuer "$CERTIFICATE_OIDC_ISSUER"');
+    expect(crypto.run).toContain("gh attestation verify");
+    expect(crypto.run).toContain("--signer-workflow");
+    expect(crypto.run).toContain('--source-ref "$EXPECTED_REF"');
+    expect(crypto.run).toContain('--source-digest "$EXPECTED_REVISION"');
+    expect(receipt.run).toContain("scripts/checks/verify-llama-cpp-image-publication-evidence.sh");
+    expect(JSON.stringify(workflow)).not.toMatch(/(?:latest|stable|release)-?alias/iu);
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply test/llama-cpp-image-workflow.test.ts]: Fixed in c3fc700f6. The workflow-contract test now collects every long-form --tag value and every action with.tags value across publication jobs, then requires the complete set to equal only $IMAGE:$CANDIDATE_TAG. This fails if any consumer or moving alias is added.

File: scripts/checks/verify-llama-cpp-image-publication-evidence.sh [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+  fi
+  jq -cS .predicate "$statement" | sha256_file /dev/stdin
+  printf '\n'
+done | sort -u >"$sbom_actual"
+if ! while IFS= read -r expected_hash; do
+  grep --fixed-strings --line-regexp "$expected_hash" "$sbom_actual" >/dev/null
+done <"$sbom_expected"; then
+  echo "ERROR: verified SBOM attestations do not match both platform documents." >&2
+  exit 1
+fi
```
[the remark this answers is not in the document]
Comment (ericksoa) [reply scripts/checks/verify-llama-cpp-image-publication-evidence.sh:311]: Fixed in c3fc700f6. The verifier now canonicalizes the expected SBOM set, requires exactly two distinct platform hashes, and uses a deterministic set-difference check so one missing platform cannot be masked by the loop exit status. Extra verified historical attestations remain permitted. The focused verifier suite includes a hostile identical-platform-SBOM case.
